alf/algorithms/agent.py

```diff
@@ -0,0 +1,266 @@
+        self._encoding_network = encoding_network
+        self._intrinsic_reward_coef = intrinsic_reward_coef
+        self._extrinsic_reward_coef = extrinsic_reward_coef
+        self._icm = intrinsic_curiosity_module
+
+    def _encode(self, time_step: ActionTimeStep):
+        observation = time_step.observation
+        if self._encoding_network is not None:
+            observation, _ = self._encoding_network(observation)
+        return observation
+
+    def predict(self, time_step: ActionTimeStep, state: AgentState):
+        """Predict for one step."""
+        observation = self._encode(time_step)
+
+        new_state = AgentState()
+
+        rl_step = self._rl_algorithm.predict(
+            time_step._replace(observation=observation), state.rl)
+        new_state = new_state._replace(rl=rl_step.state)
+
+        return PolicyStep(action=rl_step.action, state=new_state, info=())
+
+    def rollout(self,
+                time_step: ActionTimeStep,
+                state: AgentState,
+                with_experience=False):
+        """Rollout for one step."""
+        new_state = AgentState()
+        info = AgentInfo()
+        observation = self._encode(time_step)
+
+        if self._icm is not None:
+            icm_step = self._icm.train_step(
+                (observation, time_step.prev_action),
+                state=state.icm,
+                calc_intrinsic_reward=not with_experience)
+            info = info._replace(icm=icm_step.info)
+            new_state = new_state._replace(icm=icm_step.state)
+
+        rl_step = self._rl_algorithm.rollout(
+            time_step._replace(observation=observation), state.rl)
+
+        new_state = new_state._replace(rl=rl_step.state)
+        info = info._replace(rl=rl_step.info)
+
+        # TODO: can avoid computing this when collecting exps
+        if self._entropy_target_algorithm:
+            et_step = self._entropy_target_algorithm.train_step(rl_step.action)
+            info = info._replace(entropy_target=et_step.info)
+
+        return PolicyStep(action=rl_step.action, state=new_state, info=info)
+
+    def calc_training_reward(self, external_reward, info: AgentInfo):
+        """Calculate the reward actually used for training.
+
+        The training_reward includes both intrinsic reward (if there's any) and
+        the external reward.
+        Args:
+            external_reward (Tensor): reward from environment
+            info (ActorCriticInfo): (batched) policy_step.info from train_step()
+        Returns:
+            reward used for training.
+        """
+        reward = external_reward
+        if self._extrinsic_reward_coef != 1.0:
+            reward *= self._extrinsic_reward_coef
+
+        if self._icm is not None:
+            self.add_reward_summary("reward/icm", info.icm.reward)
+            reward += self._intrinsic_reward_coef * info.icm.reward
+
+        if id(reward) != id(external_reward):
```

Comment:
I think we don't need to consider that case. It does not hurt to display "overall" in the case. And it might be helpful by reminding that reward is scaled.